File: src/algorithms/graph/shortest-path/dag-shortest-path/__tests__/dag-shortest-path_test.py
import importlib
import sys
import os
import math

sys.path.insert(0, os.path.join(os.path.dirname(__file__), "..", "sources"))
module = importlib.import_module("dag-shortest-path")
dag_shortest_path = module.dag_shortest_path


def test_computes_shortest_distances_in_simple_dag():
    adj = {
        "A": [("B", 2), ("C", 6)],
        "B": [("D", 1), ("E", 4)],
        "C": [("E", 2)],
        "D": [("F", 5)],
        "E": [("F", 1)],
        "F": [],
    }
    distances = dag_shortest_path(adj, "A", ["A", "B", "C", "D", "E", "F"])
    assert distances["A"] == 0
    assert distances["B"] == 2
    assert distances["C"] == 6
    assert distances["D"] == 3
    assert distances["E"] == 6
    assert distances["F"] == 7


def test_returns_zero_distance_for_start_node():
    adj = {"Start": [("End", 5)], "End": []}
    distances = dag_shortest_path(adj, "Start", ["Start", "End"])
    assert distances["Start"] == 0


def test_returns_infinity_for_unreachable_nodes():
    adj = {"A": [("B", 3)], "B": [], "C": [("D", 2)], "D": []}
    distances = dag_shortest_path(adj, "A", ["A", "B", "C", "D"])
    assert distances["A"] == 0
    assert distances["B"] == 3
    assert math.isinf(distances["C"])
    assert math.isinf(distances["D"])


def test_handles_single_node_graph():
    adj = {"A": []}
    distances = dag_shortest_path(adj, "A", ["A"])
    assert distances["A"] == 0


def test_handles_linear_chain_correctly():
    adj = {"A": [("B", 3)], "B": [("C", 4)], "C": [("D", 2)], "D": []}
    distances = dag_shortest_path(adj, "A", ["A", "B", "C", "D"])
    assert distances["B"] == 3
    assert distances["C"] == 7
    assert distances["D"] == 9


def test_handles_negative_edge_weights_correctly():
    adj = {"A": [("B", 2), ("C", 4)], "B": [("C", -3)], "C": []}
    distances = dag_shortest_path(adj, "A", ["A", "B", "C"])
    assert distances["A"] == 0
    assert distances["B"] == 2
    assert distances["C"] == -1


def test_selects_shorter_of_two_converging_paths():
    adj = {"A": [("B", 1), ("C", 10)], "B": [("D", 2)], "C": [("D", 1)], "D": []}
    distances = dag_shortest_path(adj, "A", ["A", "B", "C", "D"])
    assert distances["D"] == 3


def test_handles_multiple_source_adjacent_nodes():
    adj = {
        "S": [("X", 1), ("Y", 4), ("Z", 2)],
        "X": [("T", 5)],
        "Y": [("T", 1)],
        "Z": [("T", 3)],
        "T": [],
    }
    distances = dag_shortest_path(adj, "S", ["S", "X", "Y", "Z", "T"])
    assert distances["T"] == 5


if __name__ == "__main__":
    test_computes_shortest_distances_in_simple_dag()
    test_returns_zero_distance_for_start_node()
    test_returns_infinity_for_unreachable_nodes()
    test_handles_single_node_graph()
    test_handles_linear_chain_correctly()
    test_handles_negative_edge_weights_correctly()
    test_selects_shorter_of_two_converging_paths()
    test_handles_multiple_source_adjacent_nodes()
    print("All tests passed!")

File: src/algorithms/graph/shortest-path/dag-shortest-path/DagShortestPath_test.java
import java.util.*;

// Compile: javac DagShortestPath.java DagShortestPath_test.java
// Run:     java -ea DagShortestPath_test
public class DagShortestPath_test {
    public static void main(String[] args) {
        testComputesShortestDistancesInSimpleDag();
        testReturnsZeroDistanceForStartNode();
        testReturnsInfinityForUnreachableNodes();
        testHandlesSingleNodeGraph();
        testHandlesLinearChainCorrectly();
        testHandlesNegativeEdgeWeightsCorrectly();
        testSelectsShorterOfTwoConvergingPaths();
        testHandlesMultipleSourceAdjacentNodes();
        System.out.println("All tests passed!");
    }

    static Map<String, List<Object[]>> adj(Object[]... entries) {
        Map<String, List<Object[]>> map = new LinkedHashMap<>();
        for (Object[] entry : entries) {
            String node = (String) entry[0];
            List<Object[]> neighbors = new ArrayList<>();
            for (int edgeIdx = 1; edgeIdx < entry.length; edgeIdx++) {
                neighbors.add((Object[]) entry[edgeIdx]);
            }
            map.put(node, neighbors);
        }
        return map;
    }

    static void testComputesShortestDistancesInSimpleDag() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"A", new Object[]{"B", 2}, new Object[]{"C", 6}},
            new Object[]{"B", new Object[]{"D", 1}, new Object[]{"E", 4}},
            new Object[]{"C", new Object[]{"E", 2}},
            new Object[]{"D", new Object[]{"F", 5}},
            new Object[]{"E", new Object[]{"F", 1}},
            new Object[]{"F"}
        );
        Map<String, Double> d = DagShortestPath.dagShortestPath(
            adjacencyList, "A", Arrays.asList("A","B","C","D","E","F"));
        assert d.get("A") == 0.0;
        assert d.get("B") == 2.0;
        assert d.get("C") == 6.0;
        assert d.get("D") == 3.0;
        assert d.get("E") == 6.0;
        assert d.get("F") == 7.0;
    }

    static void testReturnsZeroDistanceForStartNode() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"Start", new Object[]{"End", 5}},
            new Object[]{"End"}
        );
        Map<String, Double> d = DagShortestPath.dagShortestPath(
            adjacencyList, "Start", Arrays.asList("Start","End"));
        assert d.get("Start") == 0.0;
    }

    static void testReturnsInfinityForUnreachableNodes() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"A", new Object[]{"B", 3}},
            new Object[]{"B"},
            new Object[]{"C", new Object[]{"D", 2}},
            new Object[]{"D"}
        );
        Map<String, Double> d = DagShortestPath.dagShortestPath(
            adjacencyList, "A", Arrays.asList("A","B","C","D"));
        assert d.get("A") == 0.0;
        assert d.get("B") == 3.0;
        assert d.get("C") == Double.MAX_VALUE || d.get("C").isInfinite();
        assert d.get("D") == Double.MAX_VALUE || d.get("D").isInfinite();
    }

    static void testHandlesSingleNodeGraph() {
        Map<String, List<Object[]>> adjacencyList = adj(new Object[]{"A"});
        Map<String, Double> d = DagShortestPath.dagShortestPath(adjacencyList, "A", Arrays.asList("A"));
        assert d.get("A") == 0.0;
    }

    static void testHandlesLinearChainCorrectly() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"A", new Object[]{"B", 3}},
            new Object[]{"B", new Object[]{"C", 4}},
            new Object[]{"C", new Object[]{"D", 2}},
            new Object[]{"D"}
        );
        Map<String, Double> d = DagShortestPath.dagShortestPath(
            adjacencyList, "A", Arrays.asList("A","B","C","D"));
        assert d.get("B") == 3.0;
        assert d.get("C") == 7.0;
        assert d.get("D") == 9.0;
    }

    static void testHandlesNegativeEdgeWeightsCorrectly() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"A", new Object[]{"B", 2}, new Object[]{"C", 4}},
            new Object[]{"B", new Object[]{"C", -3}},
            new Object[]{"C"}
        );
        Map<String, Double> d = DagShortestPath.dagShortestPath(
            adjacencyList, "A", Arrays.asList("A","B","C"));
        assert d.get("A") == 0.0;
        assert d.get("B") == 2.0;
        assert d.get("C") == -1.0;
    }

    static void testSelectsShorterOfTwoConvergingPaths() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"A", new Object[]{"B", 1}, new Object[]{"C", 10}},
            new Object[]{"B", new Object[]{"D", 2}},
            new Object[]{"C", new Object[]{"D", 1}},
            new Object[]{"D"}
        );
        Map<String, Double> d = DagShortestPath.dagShortestPath(
            adjacencyList, "A", Arrays.asList("A","B","C","D"));
        assert d.get("D") == 3.0;
    }

    static void testHandlesMultipleSourceAdjacentNodes() {
        Map<String, List<Object[]>> adjacencyList = adj(
            new Object[]{"S", new Object[]{"X", 1}, new Object[]{"Y", 4}, new Object[]{"Z", 2}},
            new Object[]{"X", new Object[]{"T", 5}},
            new Object[]{"Y", new Object[]{"T", 1}},
            new Object[]{"Z", new Object[]{"T", 3}},
            new Object[]{"T"}
        );
        Map<String, Double> d = DagShortestPath.dagShortestPath(
            adjacencyList, "S", Arrays.asList("S","X","Y","Z","T"));
        assert d.get("T") == 5.0;
    }
}
